File: rust/periph/src/connection/connection.rs
//! Generic software-enable wrapper around an `embedded-hal` I²C bus.
//!
//! Rust never had a periph-owned struct wrapping [`I2c`] the way other
//! languages wrap their bus implementations — chip drivers here are, and
//! remain, generic directly over `embedded_hal::i2c::I2c` /
//! `embedded_hal::spi::SpiBus`. [`Connection`] is new, optional code: it adds
//! a software enable/disable gate around a generic bus for chip drivers that
//! want one, without requiring every existing I²C/SPI chip driver to change.
//!
//! For hardware EN pin control, drive an `embedded_hal::digital::OutputPin`
//! directly before constructing the chip — no wrapper is needed for that:
//!
//! ```rust,ignore
//! en_pin.set_high().unwrap();    // power up the chip
//! let conn = Connection::new(i2c);
//! let imu  = Mpu6050Minimal::new(conn);
//! ```

use embedded_hal::i2c::I2c;

/// Software-enable wrapper around a generic I²C bus.
///
/// While disabled, [`Connection::read`] zero-fills the destination buffer
/// and [`Connection::write`] is a no-op — neither touches the bus.
pub struct Connection<BUS> {
    pub(crate) bus: BUS,
    enabled: bool,
}

impl<BUS> Connection<BUS>
where
    BUS: I2c,
{
    /// Wrap `bus`, enabled by default.
    pub fn new(bus: BUS) -> Self {
        Self { bus, enabled: true }
    }

    /// Resume bus access.
    pub fn enable(&mut self) { self.enabled = true; }

    /// Gate [`Connection::read`]/[`Connection::write`] — reads zero-fill,
    /// writes become no-ops, without touching the bus.
    pub fn disable(&mut self) { self.enabled = false; }

    /// Return the current software-gate state.
    pub fn is_enabled(&self) -> bool { self.enabled }

    /// Release the underlying bus, consuming the connection.
    pub fn release(self) -> BUS {
        self.bus
    }

    /// Write `reg` then read back `buf.len()` bytes (repeated start).
    ///
    /// Zero-fills `buf` without touching the bus if disabled.
    pub(crate) fn read(&mut self, addr: u8, reg: u8, buf: &mut [u8]) -> Result<(), BUS::Error> {
        if !self.enabled {
            buf.fill(0);
            return Ok(());
        }
        self.bus.write_read(addr, &[reg], buf)
    }

    /// Write `reg` followed by `data` in one transaction.
    ///
    /// No-op if disabled.
    pub(crate) fn write(&mut self, addr: u8, reg: u8, data: &[u8]) -> Result<(), BUS::Error> {
        if !self.enabled {
            return Ok(());
        }
        let mut buf = [0u8; 17];
        buf[0] = reg;
        buf[1..=data.len()].copy_from_slice(data);
        self.bus.write(addr, &buf[..=data.len()])
    }
}

File: go/periph/connection/connection.go
// Package connection provides the abstract I2C/SPI/... bus interface used by
// every chip driver in this repository, plus a Linux and TinyGo implementation
// of the connections that drive the underlying hardware, and the InputPin /
// OutputPin abstractions used for interrupt delivery and hardware enable
// control.
//
// Selection between the Linux and TinyGo implementation of a given connection
// happens at build time via Go build tags on the implementation file
// (//go:build linux for Linux, //go:build tinygo for TinyGo). Chip drivers
// and examples import this package and reference the Connection interface
// only; only an example's main() ever names a concrete connection type.
package connection

// Connection represents one device on a bus, plus its optional INT pin, EN
// pin, and software enable gate. Each instance is bound to a single device
// address. Implementations wrap a platform-specific bus (Linux: raw ioctl()
// against /dev/i2c-N; TinyGo: machine.I2C.Tx).
type Connection interface {
	// Write sends bytes to the device.
	Write(data []byte) error
	// Read reads n bytes from the device.
	Read(n int) ([]byte, error)
	// WriteRead writes data then reads n bytes without releasing the bus
	// between the two phases (repeated start on I2C). Used for register reads.
	WriteRead(data []byte, n int) ([]byte, error)
	// Close releases the bus. After Close, the connection must not be reused.
	Close() error

	// Enable resumes bus access, driving the EN pin high if wired.
	Enable()
	// Disable gates all bus access — reads return zeros, writes are dropped
	// — and drives the EN pin low if wired.
	Disable()
	// IsEnabled returns the current software-gate state.
	IsEnabled() bool

	// IntPin returns the InputPin wired to this device's INT line, or nil
	// if none was configured.
	IntPin() InputPin
	// EnPin returns the OutputPin wired to this device's EN line, or nil
	// if none was configured.
	EnPin() OutputPin
}

// connectionBase is embedded by every concrete Connection implementation to
// share enable/disable/pin state without repeating it in each one.
type connectionBase struct {
	intPin InputPin  // nil if unused
	enPin  OutputPin // nil if unused

	disabled bool // zero value = enabled
}

// Enable resumes bus access and drives EnPin high if wired.
func (b *connectionBase) Enable() {
	b.disabled = false
	if b.enPin != nil {
		_ = b.enPin.Set(true)
	}
}

// Disable gates all bus access and drives EnPin low if wired.
func (b *connectionBase) Disable() {
	b.disabled = true
	if b.enPin != nil {
		_ = b.enPin.Set(false)
	}
}

// IsEnabled returns the current software-gate state.
func (b *connectionBase) IsEnabled() bool { return !b.disabled }

// IntPin returns the InputPin wired to this device's INT line, or nil.
func (b *connectionBase) IntPin() InputPin { return b.intPin }

// EnPin returns the OutputPin wired to this device's EN line, or nil.
func (b *connectionBase) EnPin() OutputPin { return b.enPin }
